servatrice/src/servatrice.cpp

```diff
@@ -98,14 +88,69 @@ void Servatrice_GameServer::incomingConnection(qintptr socketDescriptor)
     qDebug() << "Pool utilisation:" << debugStr;
     Servatrice_ConnectionPool *pool = connectionPools[poolIndex];
 
-    ServerSocketInterface *ssi = new ServerSocketInterface(server, pool->getDatabaseInterface());
+    TcpServerSocketInterface *ssi = new TcpServerSocketInterface(server, pool->getDatabaseInterface());
     ssi->moveToThread(pool->thread());
     pool->addClient();
     connect(ssi, SIGNAL(destroyed()), pool, SLOT(removeClient()));
 
     QMetaObject::invokeMethod(ssi, "initConnection", Qt::QueuedConnection, Q_ARG(int, socketDescriptor));
 }
 
+#if QT_VERSION > 0x050300
+#define WEBSOCKET_POOL_NUMBER 999
+
+Servatrice_WebsocketGameServer::Servatrice_WebsocketGameServer(Servatrice *_server, int _numberPools, const QSqlDatabase &_sqlDatabase, QObject *parent)
+    : QWebSocketServer("Servatrice", QWebSocketServer::NonSecureMode, parent),
+      server(_server)
+{
+    // Qt limitation: websockets can't be moved to another thread
+    Servatrice_DatabaseInterface *newDatabaseInterface = new Servatrice_DatabaseInterface(WEBSOCKET_POOL_NUMBER, server);
+    Servatrice_ConnectionPool *newPool = new Servatrice_ConnectionPool(newDatabaseInterface);
+
+    server->addDatabaseInterface(thread(), newDatabaseInterface);
+    newDatabaseInterface->initDatabase(_sqlDatabase);
+
+    connectionPools.append(newPool);
+
+    connect(this, SIGNAL(newConnection()), this, SLOT(onNewConnection()));
+}
+
+Servatrice_WebsocketGameServer::~Servatrice_WebsocketGameServer()
+{
+    for (int i = 0; i < connectionPools.size(); ++i) {
+        logger->logMessage(QString("Closing websocket pool %1...").arg(i));
+        QThread *poolThread = connectionPools[i]->thread();
+        connectionPools[i]->deleteLater(); // pool destructor calls thread()->quit()
+        poolThread->wait();
+    }
+}
+
+void Servatrice_WebsocketGameServer::onNewConnection()
+{
+   // Determine connection pool with smallest client count
+    int minClientCount = -1;
+    int poolIndex = -1;
+    QStringList debugStr;
+    for (int i = 0; i < connectionPools.size(); ++i) {
+        const int clientCount = connectionPools[i]->getClientCount();
+        if ((poolIndex == -1) || (clientCount < minClientCount)) {
+            minClientCount = clientCount;
+            poolIndex = i;
+        }
+        debugStr.append(QString::number(clientCount));
+    }
+    qDebug() << "Websocket pool utilisation:" << debugStr;
+    Servatrice_ConnectionPool *pool = connectionPools[poolIndex];
```

Comment:
Can we split out that into a helper method so we can `Servatrice_ConnectionPool *pool = findLeastUsedConnectionPool()` ?